GPII-3394: Add couchdb-prometheus-exporter

mrtyler opened the pull request:
This is an alternative approach for collecting metrics from CouchDB, suggested by @stepanstipl in https://github.com/gpii-ops/stackdriver-agent/pull/2. After some investigation and a demo today based on this PR, we decided to move forward with this solution.

Rendered README for new chart:
https://github.com/mrtyler/gpii-infra/tree/gpii-3394-cpe/shared/charts/couchdb-prometheus-exporter 

There is a commit where I only copied the gpii-preferences chart, in case you'd like to see those interdiffs. (I forgot to do the same for the gcp/modules entry and the live/dev entries; sorry.)

mrtyler commented on gcp/live/dev/k8s/gpii/couchdb-prometheus-exporter/terraform.tfvars:
```diff
@@ -0,0 +1,24 @@
+replica_count     = 1
+requests_cpu      = "25m"
+requests_memory   = "128Mi"
+limits_cpu        = "50m"
```
I remember this conversation from a different PR, but I thought we must have settled on the current scheme (memory the same, allow cpu to burst) because I copied these values from gpii-preferences.

I added a Parking Lot item to discuss this more broadly. Then, I'm happy to fix this everywhere in a follow-up PR (I noticed another small thing in gpii-preferences that I'd like to fix anyway).

stepanstipl replied:
I think we had a problem to figure the "working" values out. I still believe having same "requests" and "limits" for these types of workloads is a way to get more predictable behavior, but happy to discuss at standup.

mrtyler commented on gcp/live/dev/k8s/gpii/couchdb-prometheus-exporter/terraform.tfvars:
```diff
@@ -0,0 +1,24 @@
+
+replica_count     = 1
+requests_cpu      = "25m"
+requests_memory   = "128Mi"
```
I hid this in the commit message:

"Observed memory usage at idle and under small and moderate loads -- steady around
40MB. Tripled that for safety and rounded to 128Mi.

Observed CPU usage at idle and under small and moderate loads -- steady around 0.00 :). Rounded up to 25m for safety."

Maybe 60 or 80 is a better value? I couldn't find a good rule of thumb for this in sixty seconds of searching. I added a Parking Lot item for this discussion as well.

stepanstipl replied:
It seems to me like `3x` is very generous buffer, perhaps smth. like `1.5x` would be enough, but I don't have a good answer. What I do usually is to have a look at what the program is actually consuming and if appropriate try to put it under load to see if/how it increases and then add some small buffer on top of that.

mrtyler replied:
I settled on 80Mi, or ~2x observed memory usage.

mrtyler commented on gcp/live/prd/k8s/gpii/couchdb-prometheus-exporter/terraform.tfvars:
```diff
@@ -0,0 +1,24 @@
+
+replica_count     = 2
+requests_cpu      = "25m"
+requests_memory   = "128Mi"
```
As above.

mrtyler commented on gcp/live/prd/k8s/gpii/couchdb-prometheus-exporter/terraform.tfvars:
```diff
@@ -0,0 +1,24 @@
+
+# ↓ Module configuration (empty means all default)
+
+replica_count     = 2
```
I was thinking of [Prometheus's "High Availability" strategy](https://prometheus.io/docs/introduction/faq/#can-prometheus-be-made-highly-available):  scrape it N times and let someone downstream de-dupe it.

The incoming metrics are tagged with the Node running the metrics exporter. I think that should be sufficient if we need to de-dupe or otherwise distinguish which exporter produced a specific metric. (I think I will add some scheduling hints to prevent two exporter Pods from running on the same Node.)

The advantage of this system is simplicity: we won't lose all metrics collection during Pod or Node restarts, but the exporters don't have to know anything about one another or coordinate in any way.

stepanstipl replied:
Where will the de-duping happen?

Won't we see all the metrics twice in Stackdriver? This sounds a bit in opposition with the concern of not shipping all/too many things to Stackdriver (as we have to pay for that). I would personally wonder if it won't be too confusing seeing all the metrics 2x (if that's the case).

Otherwise, 👍 for thinking about HA, I agree that from that point of view this seems simple & good. Just need to make the de-duping happen somewhere imho.

mrtyler replied:
> This sounds a bit in opposition with the concern of not shipping all/too many things to Stackdriver (as we have to pay for that).

I don't mind paying for HA. What I don't like is paying for metrics we will definitely never use.

>  I would personally wonder if it won't be too confusing seeing all the metrics 2x (if that's the case).

In my limited experience with Stackdriver Metrics Explorer and Dashboard so far, I have not found it confusing. The incoming metrics are tagged with the Node running the metrics exporter. I think that should be sufficient if we need to de-dupe or otherwise distinguish which exporter produced a specific metric.

Do you have an alternative solution for HA that avoids these issues?

mrtyler replied:
Conversation continues [below](https://github.com/gpii-ops/gpii-infra/pull/230#pullrequestreview-184729878).

mrtyler commented on shared/charts/couchdb-prometheus-exporter/README.md:
```diff
@@ -0,0 +1,58 @@
+# CouchDB Prometheus Exporter
```
Good suggestion.

mrtyler commented on shared/charts/couchdb-prometheus-exporter/values.yaml:
```diff
@@ -0,0 +1,26 @@
+exporter_listen_port: 9984
+couchdb_uri: http://couchdb-svc-couchdb.default.svc.cluster.local:5984
+couchdb_username: admin
+couchdb_password: hunter2
```
Good suggestion.

natarajaya commented:
LGTM

mrtyler commented:
Ok, I believe I have addressed all the feedback and incorporated everything from our discussion today about cpu/memory requests/limits.

The last piece was looking into filtering the unwanted `go_*` and `process_*` metrics output by the metrics exporter. prometheus-to-sd does not have a denylist. It does have an allowlist ("whitelist") but I couldn't make it work (these logs are with `--v=9` i.e. the noisiest log level):

```
I1213 02:07:52.571009       1 translator.go:134] Exporting only whitelisted metrics: [couchdb_httpd_up couchdb_httpd_auth_cache_misses couchdb/httpd_up couchdb/httpd_auth_cache_misses custom/couchdb/httpd_up custom/couchdb/httpd_auth_cache_misses]
I1213 02:07:52.571025       1 translator.go:140] Whitelisted metric couchdb_httpd_up not present in Prometheus endpoint.
I1213 02:07:52.571030       1 translator.go:140] Whitelisted metric couchdb_httpd_auth_cache_misses not present in Prometheus endpoint.
I1213 02:07:52.571034       1 translator.go:140] Whitelisted metric couchdb/httpd_up not present in Prometheus endpoint.
I1213 02:07:52.571038       1 translator.go:140] Whitelisted metric couchdb/httpd_auth_cache_misses not present in Prometheus endpoint.
I1213 02:07:52.571042       1 translator.go:140] Whitelisted metric custom/couchdb/httpd_up not present in Prometheus endpoint.
I1213 02:07:52.571047       1 translator.go:140] Whitelisted metric custom/couchdb/httpd_auth_cache_misses not present in Prometheus endpoint.
I1213 02:07:52.571055       1 stackdriver.go:38] No metrics to send to Stackdriver for component couchdb
```

Here's some output from the exporter:

```
$ cat couchdb-prometheus-exporter.metrics.txt | grep httpd_up
# HELP couchdb_httpd_up Was the last query of CouchDB stats successful.
# TYPE couchdb_httpd_up gauge
couchdb_httpd_up 1

$ cat couchdb-prometheus-exporter.metrics.txt | grep auth_cache_miss
# HELP couchdb_httpd_auth_cache_misses number of authentication cache misses
# TYPE couchdb_httpd_auth_cache_misses gauge
couchdb_httpd_auth_cache_misses{node_name="couchdb@couchdb-couchdb-0.couchdb-couchdb.gpii.svc.cluster.local"} 134855
couchdb_httpd_auth_cache_misses{node_name="couchdb@couchdb-couchdb-1.couchdb-couchdb.gpii.svc.cluster.local"} 11429
```

I'm not sure why the names don't match. The next step would be to add debugging to prometheus-to-sd to see what it wants. However, I think this approach would force us to maintain a list of stats we want, which is undesirable.

Another approach would be to fork couchdb-prometheus-exporter and remove the unwanted stats. This may require some effort, as I think the `go_*` and `process_*` stats are coming from the Prometheus exporter framework somewhere.

Anyway, I would like to move on from this problem. I don't think the unwanted stats are a big deal for now, and our longer-term metrics collection future is somewhat uncertain. Pending new feedback, I think this PR is ready to merge.

stepanstipl commented:
@mrtyler why force-push? I can't use "see changes since last review" functionality 👎

stepanstipl commented on gcp/live/dev/k8s/gpii/couchdb-prometheus-exporter/terraform.tfvars:
```diff
@@ -0,0 +1,19 @@
+}
+
+# ↓ Module configuration (empty means all default)
+replica_count = 1
```
This is already a default value, why repeat?

stepanstipl replied:
Reading further, I noticed it's not a default in the `gcp/modules/couchdb-prometheus-exporter` module, maybe make it default there as well to avoid repetition in envs?

mrtyler replied:
Done.

stepanstipl commented on gcp/modules/couchdb-prometheus-exporter/main.tf:
```diff
@@ -0,0 +1,36 @@
+
+# Secret variables
+variable "secret_couchdb_admin_username" {}
+
```
Just cosmetic, but maybe no blank line here?

mrtyler replied:
`terraform fmt` put it there. 🤷‍♀️

stepanstipl replied:
:D ok then I guess

stepanstipl commented on shared/charts/couchdb-prometheus-exporter/README.md:
```diff
@@ -0,0 +1,58 @@
+
+Parameter | Description | Default
+--- | --- | ---
+`replicaCount` | desired number of pods | `1`
```
hm, why do we use `camelCase` here  vs `under_score` naming convention for other variables?

mrtyler replied:
I'll look at this in the follow-up cleanup PR. Looks like camel case is the convention for Charts.

stepanstipl commented on shared/charts/couchdb-prometheus-exporter/templates/deployment.yaml:
```diff
@@ -0,0 +1,62 @@
+        resources:
+{{ toYaml .Values.resources | indent 10 }}
+      - name: prometheus-to-sd
+        image: gcr.io/google-containers/prometheus-to-sd:v0.3.2
```
I think this would be pretty good to parametrize, especially the image version. Sorry I didn't notice before.

mrtyler replied:
Done.

mrtyler commented:
> why force-push? I can't use "see changes since last review" functionality

I force-pushed to make some trivial fixes in a couple commits (e.g. README formatting). I only force-pushed to new commits at the tip of the branch, never changing order or removing commits, so it is possible to reconstruct the "since last review" using the "Changes from all commits" dropdown. But that's annoying; I didn't realize Github wouldn't handle this situation. Sorry for the confusion.

mrtyler commented:
> I'm more in favor keeping more metrics than less (of course within reason), as you never know what you might need in the future for debugging some specific problem.

I agree with this sentiment. In this specific case, the stats I'd like to exclude are stats about the exporter itself -- the thing that curls a couchdb endpoint, turns the resulting json into prometheus metrics format, and returns the massaged data when curl'd. I hope we never encounter a problem where six weeks of garbage collection data from this component is the key to the solution :).

> I would probably go with just instance of exporter and relly on Kubernetes as HA mechanism and that it should restart the exporter if anything bad happens.

Counterpoint: running one instance guarantees we lose metrics on node upgrades and reboots, and on Pod re-deploys (though this is generally a smaller window). However, your arguments about de-duping are compelling so I now think starting with one instance and seeing how it does is the simplest thing.

amatas commented:
LGTM

stepanstipl commented:
👍 looks good, thanks for addressing all the bits!
